(PUP-10150) Interleave Ruby & Puppet stacks in exception reporting

justinstoller opened the pull request:
Prior to a refactor of how `PuppetStack` worked we would print the Puppet
stack frames as part of printing the Ruby stack. This was desirable in
some cases and in others it's believed that just printing the Puppet
stack would be best.

This patch re-adds the interleaving of the Puppet stack with the Ruby
stack but as an explict action during exception printing vs as a side
effect of including the Puppet stack as part of the Ruby stack.

This is achieved by adding a new accessor, `puppetstack` to our `Error`s
that deal with the reading of files. When an exception of this kind is
created it saves a reference to the current `PuppetStack`. Exceptions that
are created from these inherit the original exception's puppetstack.
`Puppet::Util::Logging.build_exception_trace` (primarily builds traces as
arrays for `ParseErrorWithIssues`),
`Puppet::Util::Logging.format_exception` (formats traces for
non-`ParseErrorWithIssues` as strings), and their underlying usage of
`Puppet::Util.pretty_backtrace` have been refactored to honor the
puppetstack if configured and available on the exception in question.

It also adds a new setting `:puppet_trace` which will only print the
Puppet stack, without any of the Ruby stack frames for those authors
working solely on Puppet code.

justinstoller commented on lib/puppet/pops/issue_reporter.rb:
```diff
@@ -106,11 +102,30 @@ def self.create_exception(exception_class, emit_message, formatter, diagnostic)
+    end
+
+    if diagnostic.exception && diagnostic.exception.respond_to?(:puppet_stacktrace)
+      exception.puppet_stacktrace = diagnostic.exception.puppet_stacktrace
```
just before I pushed this up I noticed this had a typo, so it must not be necessary for the case I was testing.

need to figure out a way to verify that we don't have to do shenanigans like this at various other points in our error handling code - we like to repackage exceptions.

puppetcla commented:
CLA signed by all contributors.

Magisus commented:
This is targeted at master, I think it should be 5.5.x?

justinstoller commented:
> This is targeted at master, I think it should be 5.5.x?

Good catch, miss-targeted the PR. Updated.

Magisus commented:
Fixing hooks

justinstoller commented:
Rebased and updated the commit message / pr body. I think this is ready to go now.

Magisus commented on lib/puppet/defaults.rb:
```diff
@@ -186,6 +186,11 @@ def self.valid_file_checksum_types
         end
     },
+    :puppet_trace => {
+        :default  => false,
```
Does having this default to false mimic previous behavior, or maintain the performance improvement from the previous patch?

justinstoller replied:
It doesn't have an effect on either, actually. Right now the PR will interleave the ruby and puppet stack traces if the `--trace` flag is passed (rationale being the expensive part is the getting of the Ruby trace, adding the puppet trace is relatively inexpensive and doesn't dilute the signal to noise ratio much if at all), if the `--trace` flag isn't passed but the `--puppet_trace` flag _is_, then we will just output the puppet stack trace, which everyone thought was the most value for the majority of users (high signal to noise ratio for actual puppet authors). ie, using the reproducer in the ticket,
with `--trace`:

```
Error: Evaluation Error: Error while evaluating a Function Call, somethinng (file: /Users/justin/Backup/code/work/puppet/broken.pp, line: 2, column: 3) on node localhost
/Users/justin/Backup/code/work/puppet/lib/puppet/parser/functions/fail.rb:3:in block in <top (required)>'
/Users/justin/Backup/code/work/puppet/lib/puppet/parser/functions.rb:174:in block (2 levels) in newfunction'
/Users/justin/Backup/code/work/puppet/lib/puppet/util/profiler/around_profiler.rb:58:in profile'
/Users/justin/Backup/code/work/puppet/lib/puppet/util/profiler.rb:51:in profile'
/Users/justin/Backup/code/work/puppet/lib/puppet/parser/functions.rb:167:in block in newfunction'
/Users/justin/Backup/code/work/puppet/broken.pp:2
/Users/justin/Backup/code/work/puppet/lib/puppet/pops/puppet_stack.rb:41:in stack'
...

```





with just `--puppet_trace`:

```
Error: Evaluation Error: Error while evaluating a Function Call, somethinng (file: /Users/justin/Backup/code/work/puppet/broken.pp, line: 2, column: 3) on node localhost
/Users/justin/Backup/code/work/puppet/broken.pp:2
/Users/justin/Backup/code/work/puppet/broken.pp:6
/Users/justin/Backup/code/work/puppet/broken.pp:9
/Users/justin/Backup/code/work/puppet/broken.pp:0
```

Magisus replied:
Ahh I get it, thanks for clearing that up.

Magisus replied:
Think you could add something to the doc string for `trace` noting that it will contain both the Ruby and the Puppet frames? I think that would help disambiguate these two.

justinstoller replied:
Updated, let me know if it's still lacking.

mwaggett replied:
What happens if both `--trace` and `--puppet_trace` are passed?

justinstoller replied:
`--puppet_trace` is a noop then, I can add a comment about that as well!

justinstoller replied:
updated

mwaggett commented on spec/unit/util/log_spec.rb:
```diff
@@ -256,9 +289,108 @@ def log_notice(message)
+      expect(logs.size).to eq(1)
+      log = logs[0]
+      expect(log.message).to_not match('/log_spec.rb')
+      expect(log.backtrace[0]).to match('/log_spec.rb')
```
nitpick, but above you use `.first` instead of `[0]`, if you wanna be consistent

mwaggett commented on spec/unit/util/log_spec.rb:
```diff
@@ -256,9 +289,108 @@ def log_notice(message)
+      Puppet[:trace] = false
+      Puppet[:puppet_trace] = false
+
+      # Normal logging exepectations are still true
```
```suggestion
      # Normal logging expectations are still true
```

mwaggett replied:
this is in a bunch of places.. 😅

mwaggett commented:
This looks sane to me, but I'm curious how you tested (besides the spec tests, which I found somewhat hard to follow in terms of what all the different test cases are actually checking).

justinstoller commented:
Glen provides a reproducer manifest in the ticket. I used this with ‘bundle exec puppet apply broken.pp <trace flags>’

joshcooper commented on spec/unit/util/log_spec.rb:
```diff
@@ -256,9 +289,108 @@ def log_notice(message)
+    it "backtrace has interleaved PuppetStack when logging ParseErrorWithIssue" do
+      logs = []
+      destination = Puppet::Test::LogCollector.new(logs)
+      Puppet::Util::Log.newdestination(destination)
```
puppet's spec_helper creates a `LogCollector` and makes it available to each rspec example as `@logs` instance variable. So I think you can drop the part above, and below do `expect(@logs.size).to eq(1)`

justinstoller replied:
Oh, nice!

justinstoller commented:
Addressed Molly's spelling and consistency comments & reduced the test code using the helpers Josh pointed out. I also moved the tests from log_spec to logging_spec since they're all testing the log_exception function provided by logging. I also added some describe blocks and reworded some test titles to try and make them easier to follow. Let me know if there's more I should do!

justinstoller commented:
(just pushed a fix to a typo in the first line of the commit message "how how" -> "of how")
